torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,55 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType typeToScalarType(const c10::TypePtr & type) {
+    if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    }
+    return scalarTypeFromJitType(type);
+  }
+
+  c10::ScalarType getPromotedTypeForArithmeticOp(Node *node) {
+    c10::ScalarType dimmed = c10::ScalarType::Undefined;
+    c10::ScalarType zerodim = c10::ScalarType::Undefined;
+    auto& args = node->schema().arguments();
+    for (size_t i = 0 ; i < args.size() ; i++ ) {
+      auto arg = args[i];
+      auto dtt = node->inputs()[i]->type()->cast<DimensionedTensorType>();
+      if ( dtt && dtt->dim() > 0) {
+        if (dtt->dim() > 0) {
+          if (dimmed == c10::ScalarType::Undefined) {
+            dimmed = dtt->scalarType();
+          } else {
+            dimmed = c10::promoteTypes(dimmed, dtt->scalarType());
+          }
+        }
+      } else {
+        auto inputDtype = typeToScalarType(node->inputs()[i]->type());
+        if (zerodim == c10::ScalarType::Undefined) {
```

Comment:
This case isn't possible

Reply:
why not? I think zerodim will be undefined on the first pass?